File: Source/TWXProxy-C#/Log.cs
using System;
using System.IO;
using System.Windows.Forms;
using Core;
using Global;
using Ansi;
using Utility;
 // 
 // Copyright (C) 2005  Remco Mulder
 // 
 // This program is free software; you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
 // the Free Software Foundation; either version 2 of the License, or
 // (at your option) any later version.
 // 
 // This program is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.
 // 
 // You should have received a copy of the GNU General Public License
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 // 
 // For source notes please refer to Notes.txt
 // For license terms please refer to GPL.txt.
 // 
 // These files should be stored in the root of the compression you
 // received this source in.
 // This unit is responsible for logging data recieved from the server.
namespace Log
{
    public struct TLogEntry
    {
        public uint Timestamp;
        public ushort EntrySize;
        public byte Entry;
    } // end TLogEntry

    public class TModLog: TTWXModule, IModLog, ITWXGlobals
    {
        protected Timer LogTimer
        {
          get {
            return GetLogTimer();
          }
        }
        public string LogFilename
        {
          get {
            return FLogFilename;
          }
        }
        public bool LogFileOpen
        {
          get {
            return FLogFileOpen;
          }
        }
        public bool PlayingLog
        {
          get {
            return FPlayingLog;
          }
        }
        public bool BinaryLogs
        {
          get {
            return GetBinaryLogs();
          }
          set {
            SetBinaryLogs(value);
          }
        }
        public bool LogANSI
        {
          get {
            return GetLogANSI();
          }
          set {
            SetLogANSI(value);
          }
        }
        public bool LogData
        {
          get {
            return GetLogData();
          }
          set {
            SetLogData(value);
          }
        }
        public uint MaxPlayDelay
        {
          get {
            return FMaxPlayDelay;
          }
          set {
            FMaxPlayDelay = value;
          }
        }
        public bool NotifyPlayCuts
        {
          get {
            return FNotifyPlayCuts;
          }
          set {
            FNotifyPlayCuts = value;
          }
        }
        private string FProgramDir = String.Empty;
        private string FLogFilename = String.Empty;
        private bool FLogFileOpen = false;
        private System.IO.File FLogFile = null;
        private System.IO.File FPlayLogFile = null;
        private DateTime FLastLog;
        private bool FLogData = false;
        private bool FLogANSI = false;
        private bool FBinaryLogs = false;
        private bool FPlayingLog = false;
        private Timer FLogTimer = null;
        private TLogEntry FNextLogEntry = null;
        // pointer to next log entry to be played
        private uint FLastPlayTime = 0;
        private uint FMaxPlayDelay = 0;
        private bool FNotifyPlayCuts = false;
        public TTWXModule(object AOwner, IPersistenceController APersistenceController) : base(AOwner, APersistenceController)
        {
        }
        public override void AfterConstruction()
        {
            base.AfterConstruction();
            // set defaults
            FNotifyPlayCuts = true;
            FMaxPlayDelay = 10000;
        }

        public override void BeforeDestruction()
        {
            // ensure log file is closed
            CloseLog();
            //@ Unsupported property or method(C): 'Free'
            FLogTimer.Free;
            if ((FNextLogEntry != null))
            {
                //@ Unsupported function or procedure: 'FreeMem'
                FreeMem(FNextLogEntry);
            }
            base.BeforeDestruction();
        }

        private void LogTimerTimer(Object Sender)
        {
            string Text;
            // show the cached entry, then cache the next one
            //@ Unsupported function or procedure: 'SetString'
            SetString(Text, ((FNextLogEntry.Entry) as string), FNextLogEntry.EntrySize);
            Global.Units.Global.TWXServer.Broadcast(Text);
            //@ Unsupported function or procedure: 'FreeMem'
            FreeMem(FNextLogEntry);
            FNextLogEntry = GetNextLogEntry();
        }

        private TLogEntry GetNextLogEntry()
        {
            TLogEntry result;
            object Buf;
            uint X;
            if ((FPlayLogFile.BaseStream.Position + sizeof(TLogEntry) >= FPlayLogFile.BaseStream.Length))
            {
                // At end of log file
                EndPlayLog();
                result = null;
            }
            else
            {
                // Read the entry header
                //@ Unsupported function or procedure: 'AllocMem'
                result = AllocMem(sizeof(TLogEntry));
                //@ Unsupported function or procedure: 'BlockRead'
                BlockRead(FPlayLogFile, result, sizeof(TLogEntry) - 1);
                // Read the rest of the entry now that we know its size
                if ((result.EntrySize == 0))
                {
                    // This is a dud frame.  Skip straight to the next record
                    //@ Unsupported function or procedure: 'FreeMem'
                    FreeMem(result);
                    result = result;
                    LogTimer.Interval = 1;
                }
                else
                {
                    //@ Unsupported function or procedure: 'ReAllocMem'
                    ReAllocMem(result, sizeof(TLogEntry) + result.EntrySize - 1);
                    Buf = (result + sizeof(TLogEntry) - 1 as object);
                    //@ Unsupported function or procedure: 'BlockRead'
                    BlockRead(FPlayLogFile, Buf, result.EntrySize);
                    // Set the timer to play it
                    if ((FLastPlayTime == 0))
                    {
                        LogTimer.Interval = 1;
                    }
                    else
                    {
                        if ((FLastPlayTime == result.Timestamp))
                        {
                            X = 1;
                        }
                        else if ((FLastPlayTime > result.Timestamp))
                        {
                            //@ Unsupported function: 'High'
                            X = result.Timestamp + (High(X) - FLastPlayTime);
                        }
                        else
                        {
                            X = result.Timestamp - FLastPlayTime;
                        }
                        if ((X > MaxPlayDelay))
                        {
                            LogTimer.Interval = 5000;
                            if (NotifyPlayCuts)
                            {
                                Global.Units.Global.TWXServer.ClientMessage("PLAYBACK: Long delay of " + (X / 1000).ToString() + "s cut to 5s");
                            }
                        }
                        else
                        {
                            LogTimer.Interval = X;
                        }
                    }
                    FLastPlayTime = result.Timestamp;
                }
                LogTimer.Enabled = true;
            }
            return result;
        }

        public void BeginPlayLog(string Filename)
        {
            if (PlayingLog)
            {
                return;
            }
            Environment.CurrentDirectory = FProgramDir;
            FPlayLogFile = new FileInfo(Filename);
            StreamReader _R_2 = FPlayLogFile.OpenText();
            //@ Unsupported function or procedure: 'IOResult'
            if ((IOResult != 0))
            {
                Global.Units.Global.TWXServer.ClientMessage("Unable to open capture file \'" + LogFilename + "\' for play.");
            }
            else
            {
                Global.Units.Global.TWXServer.ClientMessage("Beginning playback of capture file: " + Ansi.Units.Ansi.ANSI_7 + Filename + Core.Units.Core.endl + Ansi.Units.Ansi.ANSI_15 + "Press any key to terminate.");
                FPlayingLog = true;
                if ((FNextLogEntry != null))
                {
                    //@ Unsupported function or procedure: 'FreeMem'
                    FreeMem(FNextLogEntry);
                }
                FNextLogEntry = GetNextLogEntry();
            // starts log timer
            }
        }

        public void EndPlayLog()
        {
            if (FPlayingLog)
            {
                FPlayingLog = false;
                FLastPlayTime = 0;
                LogTimer.Enabled = false;
                _R_2.Close();
                Global.Units.Global.TWXServer.ClientMessage("Playback of capture file completed.");
            }
        }

        private string GetLogName()
        {
            string result;
            result = "logs\\" + (DateTime.Today).ToString() + ' ' + Utility.Units.Utility.StripFileExtension(Utility.Units.Utility.ShortFilename(Global.Units.Global.TWXDatabase.DatabaseName));
            Utility.Units.Utility.Replace(ref result, '/', '-');
            Utility.Units.Utility.Replace(ref result, ':', '-');
            if (FBinaryLogs)
            {
                result = result + ".cap";
            }
            else
            {
                result = result + ".log";
            }
            return result;
        }

        private Timer GetLogTimer()
        {
            Timer result;
            if (!(FLogTimer != null))
            {
                FLogTimer = new Timer(this);
                FLogTimer.Tick = LogTimerTimer;
                FLogTimer.Interval = 1;
                FLogTimer.Enabled = false;
            }
            result = FLogTimer;
            return result;
        }

        private void OpenLog(string Filename)
        {
            if (LogFileOpen || !(Global.Units.Global.TWXDatabase.DataBaseOpen))
            {
                return;
            }
            Environment.CurrentDirectory = FProgramDir;
            FLogFilename = Filename;
            FLogFile = new FileInfo(LogFilename);
            StreamReader _R_3 = FLogFile.OpenText();
            //@ Unsupported function or procedure: 'IOResult'
            if ((IOResult != 0))
            {
                StreamWriter _W_1 = FLogFile.CreateText();
            }
            //@ Unsupported function or procedure: 'IOResult'
            if ((IOResult != 0))
            {
                MessageBox.Show("Unable to open log file - logging has been disabled." + Core.Units.Core.endl + "File: " + LogFilename);
                LogData = false;
            }
            else
            {
                FLogFileOpen = true;
                FLastLog = DateTime.Now;
            }
        }

        private void CloseLog()
        {
            if (!LogFileOpen)
            {
                return;
            }
            _W_1.Close();
            FLogFileOpen = false;
        }

        private void LogBinary(string Data)
        {
            TLogEntry LogEntry;
            int Size;
            // log timestamp along with packet
            Size = sizeof(TLogEntry) + Data.Length - 1;
            //@ Unsupported function or procedure: 'AllocMem'
            LogEntry = AllocMem(Size);
            //@ Unsupported function or procedure: 'GetTickCount'
            LogEntry.Timestamp = GetTickCount;
            LogEntry.EntrySize = Data.Length;
            //@ Undeclared identifier(3): 'CopyMemory'
            CopyMemory((LogEntry.Entry), (Data as object), Data.Length);
            //@ Unsupported function or procedure: 'BlockWrite'
            BlockWrite(FLogFile, LogEntry, Size);
            //@ Unsupported function or procedure: 'FreeMem'
            FreeMem(LogEntry);
        }

        public void DoLogData(string Data, string ANSIData)
        {
            int I;
            if (LogData && LogFileOpen)
            {
                if (!(Global.Units.Global.TWXInterpreter.LogData))
                {
                    return;
                }
                if ((Convert.ToInt64(DateTime.Now) != Convert.ToInt64(FLastLog)))
                {
                    // next day - open the next log
                    CloseLog();
                    OpenLog(GetLogName());
                }
                _R_3.BaseStream.Seek(_W_1.BaseStream.Length, SeekOrigin.Begin);
                if (FBinaryLogs)
                {
                    if (LogANSI)
                    {
                        LogBinary(ANSIData);
                    }
                    else
                    {
                        LogBinary(Data);
                    }
                }
                else
                {
                    if (LogANSI)
                    {
                        //@ Unsupported function or procedure: 'BlockWrite'
                        BlockWrite(FLogFile, (ANSIData as string), ANSIData.Length, I);
                    }
                    else
                    {
                        //@ Unsupported function or procedure: 'BlockWrite'
                        BlockWrite(FLogFile, (Data as string), Data.Length, I);
                    }
                }
                //@ Unsupported function or procedure: 'IOResult'
                if ((IOResult != 0))
                {
                    MessageBox.Show("TWX Proxy has encountered an error logging data sent from the server.  " + Core.Units.Core.endl + "This could be due to insufficient disk space or the log file is in use.  Logging has been disabled.");
                    LogData = false;
                }
            }
        }

        public void DatabaseChanged()
        {
            string Filename;
            if (LogData)
            {
                Filename = GetLogName();
                if ((Filename != FLogFilename))
                {
                    CloseLog();
                    OpenLog(Filename);
                }
            }
        }

        // IModLog
        protected bool GetLogData()
        {
            bool result;
            result = FLogData;
            return result;
        }

        protected void SetLogData(bool Value)
        {
            if ((FLogData != Value))
            {
                FLogData = Value;
                if (Value)
                {
                    OpenLog(GetLogName());
                }
                else
                {
                    CloseLog();
                }
            }
        }

        protected bool GetLogANSI()
        {
            bool result;
            result = FLogANSI;
            return result;
        }

        protected void SetLogANSI(bool Value)
        {
            FLogANSI = Value;
        }

        protected bool GetBinaryLogs()
        {
            bool result;
            result = FBinaryLogs;
            return result;
        }

        protected void SetBinaryLogs(bool Value)
        {
            FBinaryLogs = Value;
        }

        // ITWXGlobals
        protected void SetProgramDir(string Value)
        {
            FProgramDir = Value;
        }

        protected string GetProgramDir()
        {
            string result;
            result = FProgramDir;
            return result;
        }

    } // end TModLog

}

File: twxp/Source/TWX26/Log.pas
{
Copyright (C) 2005  Remco Mulder

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For source notes please refer to Notes.txt
For license terms please refer to GPL.txt.

These files should be stored in the root of the compression you 
received this source in.
}
// This unit is responsible for logging data recieved from the server.

unit
  Log;

interface

uses
  SysUtils,
  Classes,
  ExtCtrls,
  Core;

type
  PLogEntry = ^TLogEntry;
  TLogEntry = packed record
    Timestamp: Cardinal;
    EntrySize: Word;
    Entry: Byte; // Implicit entry stored at end of record (defined by dynamically allocated size)
  end;

  TModLog = class(TTWXModule, IModLog, ITWXGlobals)
  private
    FProgramDir     : string;
    FLogFilename    : string;
    FLogFileOpen    : Boolean;
    FLogFile        : file;
    FPlayLogFile    : file;
    FLastLog        : TDateTime;
    FLogEnabled     : Boolean;
    FLogData        : Boolean;
    FLogANSI        : Boolean;
    FBinaryLogs     : Boolean;
    FPlayingLog     : Boolean;
    FLogTimer       : TTimer;
    FNextLogEntry   : PLogEntry; // pointer to next log entry to be played
    FLastPlayTime   : Cardinal;
    FMaxPlayDelay   : Cardinal;
    FNotifyPlayCuts : Boolean;

    function GetLogName: string;
    function GetLogTimer: TTimer;
    function GetNextLogEntry: PLogEntry;
    procedure OpenLog(Filename: string);
    procedure CloseLog;
    procedure LogBinary(const Data: string);
    procedure LogTimerTimer(Sender: TObject);

  protected
    { ITWXGlobals }
    procedure SetProgramDir(const Value: string);
    function GetProgramDir: string;

    { IModLog }
    function GetLogEnabled: Boolean;
    procedure SetLogEnabled(Value: Boolean);
    function GetLogData: Boolean;
    procedure SetLogData(Value: Boolean);
    function GetLogANSI: Boolean;
    procedure SetLogANSI(Value: Boolean);
    function GetBinaryLogs: Boolean;
    procedure SetBinaryLogs(Value: Boolean);

    property LogTimer: TTimer read GetLogTimer;
  public
    procedure AfterConstruction; override;
    procedure BeforeDestruction; override;
    procedure WriteLog(const Data: string);
    procedure DoLogData(const Data: string; const ANSIData: string);
    procedure DatabaseChanged;
    procedure BeginPlayLog(const Filename: string);
    procedure EndPlayLog;

    property LogFilename : string read FLogFilename;
    property LogFileOpen : Boolean read FLogFileOpen;
    property PlayingLog: Boolean read FPlayingLog;
  published
    property BinaryLogs: Boolean read GetBinaryLogs write SetBinaryLogs;
    property LogANSI: Boolean read GetLogANSI write SetLogANSI;
    property LogData: Boolean read GetLogData write SetLogData;
    property LogEnabled: Boolean read GetLogEnabled write SetLogEnabled;
    property MaxPlayDelay: Cardinal read FMaxPlayDelay write FMaxPlayDelay;
    property NotifyPlayCuts: Boolean read FNotifyPlayCuts write FNotifyPlayCuts;
  end;

implementation

uses
  Global,
  ANSI,
  Utility,
  Dialogs,
  Windows;

procedure TModLog.AfterConstruction;
begin
  inherited;

  // set defaults
  FLogEnabled := True;
  FLogData := True;
  FNotifyPlayCuts := True;
  FMaxPlayDelay := 10000;
end;

procedure TModLog.BeforeDestruction;
begin
  // ensure log file is closed
  CloseLog;
  
  FLogTimer.Free;

  if Assigned(FNextLogEntry) then
    FreeMem(FNextLogEntry);

  inherited;
end;

procedure TModLog.LogTimerTimer(Sender: TObject);
var
  Text: string;
begin
  // show the cached entry, then cache the next one
  SetString(Text, PChar(@(FNextLogEntry.Entry)), FNextLogEntry.EntrySize);
  TWXServer.Broadcast(Text);
  FreeMem(FNextLogEntry);
  FNextLogEntry := GetNextLogEntry;
end;

function TModLog.GetNextLogEntry: PLogEntry;
var
  Buf: Pointer;
  X: Cardinal;
begin
  if (FilePos(FPlayLogFile) + SizeOf(TLogEntry) >= FileSize(FPlayLogFile)) then
  begin
    // At end of log file
    EndPlayLog;
    Result := nil;
  end
  else
  begin
    // Read the entry header
    Result := AllocMem(SizeOf(TLogEntry));
    BlockRead(FPlayLogFile, Result^, SizeOf(TLogEntry) - 1);

    // Read the rest of the entry now that we know its size
    if (Result.EntrySize = 0) then
    begin
      // This is a dud frame.  Skip straight to the next record
      FreeMem(Result);
      Result := GetNextLogEntry;
      LogTimer.Interval := 1;
    end
    else
    begin
      ReAllocMem(Result, SizeOf(TLogEntry) + Result.EntrySize - 1);
      Buf := Pointer(Integer(Result) + SizeOf(TLogEntry) - 1);
      BlockRead(FPlayLogFile, Buf^, Result.EntrySize);

      // Set the timer to play it
      if (FLastPlayTime = 0) then
        LogTimer.Interval := 1
      else
      begin
        if (FLastPlayTime = Result.Timestamp) then
          X := 1
        else if (FLastPlayTime > Result.Timestamp) then
          X := Result.Timestamp + (High(X) - FLastPlayTime)
        else
          X := Result.Timestamp - FLastPlayTime;

        if (X > MaxPlayDelay) then begin
          LogTimer.Interval := 5000;

          if NotifyPlayCuts then
            TWXServer.ClientMessage('PLAYBACK: Long delay of ' + IntToStr(X div 1000) + 's cut to 5s');
        end else
          LogTimer.Interval := X;
      end;

      FLastPlayTime := Result.Timestamp;
    end;

    LogTimer.Enabled := True;
  end;
end;

procedure TModLog.BeginPlayLog(const Filename: string);
begin
  if (PlayingLog) then
    Exit;

  SetCurrentDir(FProgramDir);

  AssignFile(FPlayLogFile, Filename);
  {$I-}
  Reset(FPlayLogFile, 1);

  if (IOResult <> 0) then
    TWXServer.ClientMessage('Unable to open capture file ''' + LogFileName + ''' for play.')
  else
  begin
    TWXServer.ClientMessage('Beginning playback of capture file: ' + ANSI_7 + Filename + endl + ANSI_15 + 'Press any key to terminate.');
    FPlayingLog := True;

    if Assigned(FNextLogEntry) then
      FreeMem(FNextLogEntry);

    FNextLogEntry := GetNextLogEntry; // starts log timer
  end;
  {$I+}
end;

procedure TModLog.EndPlayLog;
begin
  if (FPlayingLog) then begin
    FPlayingLog := False;
    FLastPlayTime := 0;
    LogTimer.Enabled := False;
    CloseFile(FPlayLogFile);
    TWXServer.ClientMessage('Playback of capture file completed.');
  end;
end;

function TModLog.GetLogName: string;
begin
  Result := 'logs\' + DateToStr(Date) + ' ' + StripFileExtension(ShortFileName(TWXDatabase.DatabaseName));
  Replace(Result, '/', '-');
  Replace(Result, ':', '-');

  if FBinaryLogs then
    Result := Result + '.cap'
  else
    Result := Result + '.log';
end;

function TModLog.GetLogTimer: TTimer;
begin
  if not Assigned(FLogTimer) then begin
    FLogTimer := TTimer.Create(Self);

    with FLogTimer do
    begin
      OnTimer := LogTimerTimer;
      Interval := 1;
      Enabled := False;
    end;
  end;

  Result := FLogTimer;
end;

procedure TModLog.OpenLog(Filename: string);
begin
  if (LogFileOpen) or not (TWXDatabase.DatabaseOpen) then
    Exit;

  SetCurrentDir(FProgramDir);
  FLogFileName := Filename;

  AssignFile(FLogFile, LogFileName);
  {$I-}
  Reset(FLogFile, 1);

  if (IOResult <> 0) then
    ReWrite(FLogFile, 1);

  // MB - Ignore reslut 2 ??? caused by mombot.
  if (IOResult <> 0) and (IOResult <> 2) then
  begin
    TWXServer.ClientMessage('Unable to open log file - logging has been disabled. ' + endl +
                             'File: ' + LogFileName + '        IOResult: = ' + IntToStr(IOResult));
    LogData := FALSE;
  end
  else
  begin
    LogData := TRUE;
    FLogFileOpen := TRUE;
    FLastLog := Now;
  end;
  {$I+}
end;

procedure TModLog.CloseLog;
begin
  if not (LogFileOpen) then
    Exit;

  {$I-}
  CloseFile(FLogFile);
  FLogFileOpen := FALSE;
  {$I+}
end;

procedure TModLog.LogBinary(const Data: string);
var
  LogEntry: PLogEntry;
  Size: Integer;
begin
  // log timestamp along with packet
  Size := SizeOf(TLogEntry) + Length(Data) - 1;
  LogEntry := AllocMem(Size);
  LogEntry.Timestamp := GetTickCount;
  LogEntry.EntrySize := Length(Data);
  CopyMemory(@(LogEntry^.Entry), Pointer(Data), Length(Data));
  BlockWrite(FLogFile, LogEntry^, Size);
  FreeMem(LogEntry);
end;

procedure TModLog.WriteLog(const Data: string);
begin
  if LogEnabled then
  begin
    // Make sure the log file is open.
    if not LogFileOpen then
      OpenLog(GetLogName)
    else if (Trunc(Now) <> Trunc(FLastLog)) then
    begin
      // next day - open the next log
      CloseLog;
      OpenLog(GetLogName);
    end;

    {$I-}
    Seek(FLogFile, FileSize(FLogFile));
    BlockWrite(FLogFile, PChar(Data)^, Length(Data));
    {$I+}

  end;
end;


procedure TModLog.DoLogData(const Data: string; const ANSIData: string);
var
  I: Integer;
begin
  if LogData and LogEnabled then
  begin
    // Making sure the log file is open.
    if not LogFileOpen then
      OpenLog(GetLogName)
    else if (Trunc(Now) <> Trunc(FLastLog)) then
    begin
      // next day - open the next log
      CloseLog;
      OpenLog(GetLogName);
    end;

    {$I-}
    Seek(FLogFile, FileSize(FLogFile));

    if (FBinaryLogs) then
    begin
      if (LogANSI) then
        LogBinary(ANSIData)
      else
        LogBinary(Data);
    end
    else
    begin
      if (LogANSI) then
        BlockWrite(FLogFile, PChar(ANSIData)^, Length(ANSIData), I)
      else
        BlockWrite(FLogFile, PChar(Data)^, Length(Data), I);
    end;

    if (IOResult <> 0) then
    begin
      // MB - This error seems to be non-fatal - We may need to re-visit this, but for now I am going to ignore it.
      // TWXServer.ClientMessage('TWX Proxy has encountered an error logging data sent from the server.  ' + endl + 'This could be due to insufficient disk space or the log file is in use.  Logging has been disabled.');
      // LogData := FALSE;
    end;

    {$I+}
  end;
end;

procedure TModLog.DatabaseChanged;
var
  Filename: string;
begin
  if (LogEnabled) then
  begin
    Filename := GetLogName;

    if (Filename <> FLogFileName) then
    begin
      CloseLog;
      OpenLog(Filename);
    end;
  end;
end;

function TModLog.GetLogData: Boolean;
begin
  Result := FLogData;
end;

procedure TModLog.SetLogData(Value: Boolean);
begin
  FLogData := Value;
end;

function TModLog.GetLogEnabled: Boolean;
begin
  Result := FLogEnabled;
end;

procedure TModLog.SetLogEnabled(Value: Boolean);
begin
  if (FLogEnabled <> Value) then
  begin
    FLogEnabled := Value;

    if (Value) then
      OpenLog(GetLogName)
    else
      CloseLog;
  end;
end;

function TModLog.GetLogANSI: Boolean;
begin
  Result := FLogANSI;
end;

procedure TModLog.SetLogANSI(Value: Boolean);
begin
  FLogANSI := Value;
end;

function TModLog.GetBinaryLogs: Boolean;
begin
  Result := FBinaryLogs;
end;

procedure TModLog.SetBinaryLogs(Value: Boolean);
begin
  FBinaryLogs := Value;
end;

procedure TModLog.SetProgramDir(const Value: string);
begin
  FProgramDir := Value;
end;

function TModLog.GetProgramDir: string;
begin
  Result := FProgramDir;
end;

end.
